File: Plugins/Orimath.Basics.View/ViewModels/EdgeViewModel.cs
using System;
using Orimath.Controls;
using Orimath.Core;
using Orimath.Plugins;

namespace Orimath.Basics.View.ViewModels
{
    public class EdgeViewModel : NotifyPropertyChanged, IDisplayTargetViewModel
    {
        private readonly ILayerModel? _layer;
        public EdgeViewModel(ILayerModel? layer, Edge edge, IViewPointConverter pointConverter)
        {
            _layer = layer;
            Source = edge;
            (X1, Y1) = pointConverter.ModelToView(edge.Line.Point1);
            (X2, Y2) = pointConverter.ModelToView(edge.Line.Point2);
        }

        public Edge Source { get; }
        public double X1 { get; }
        public double Y1 { get; }
        public double X2 { get; }
        public double Y2 { get; }
        public double XFactor => Source.Line.Line.XFactor;
        public double YFactor => NearlyEquatable.Negate(Source.Line.Line.YFactor);
        public double Intercept => Source.Line.Line.Intercept;
        public double Slope => NearlyEquatable.UnaryPlus(XFactor / YFactor);
        public double Angle => NearlyEquatable.UnaryPlus(Math.Atan2(XFactor, YFactor) / Math.PI * 180.0) % 180.0;
        public double Length => Source.Line.Length;

        public (ILayerModel, DisplayTarget) GetTarget()
        {
            if (_layer is null) throw new InvalidOperationException();
            return (_layer, DisplayTarget.NewEdge(Source));
        }
        public override string ToString() => $"{Source.Line.Line}\r\n傾き:{Slope:0.#####} 角度:{Angle:0.#####}°";
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/EdgeViewModel.fs
namespace Orimath.Basics.View.ViewModels
open System
open Orimath.Core
open Orimath.Plugins
open Orimath.Controls
open Orimath.Basics.View
open Orimath.Core.NearlyEquatable

type EdgeViewModel(edge: Edge, pointConverter: IViewPointConverter, ?layer: ILayerModel) =
    inherit NotifyPropertyChanged()
    let (ScreenPoint(x1, y1)) = pointConverter.ModelToView(edge.Point1)
    let (ScreenPoint(x2, y2)) = pointConverter.ModelToView(edge.Point2)
    member _.Source = edge
    member _.X1 = x1
    member _.Y1 = y1
    member _.X2 = x2
    member _.Y2 = y2
    member _.XFactor = edge.Line.XFactor
    member _.YFactor = !-edge.Line.YFactor
    member _.Intercept = edge.Line.Intercept
    member _.Slope = !+(edge.Line.XFactor / -edge.Line.YFactor)
    member _.Angle = !+(atan2 edge.Line.XFactor -edge.Line.YFactor / Math.PI * 180.0) % 180.0
    member _.Length = edge.Length
   
    override _.ToString() = edge.ToString()

    interface IDisplayTargetViewModel with
        member _.GetTarget() =
            match layer with
            | Some(l) -> l, DisplayTarget.Edge(edge)
            | None -> invalidOp "Layer is not set"
